tokio/src/time/driver/entry.rs

```diff
@@ -1,361 +1,747 @@
+    /// take the waker (as an acq/rel barrier) and arrange to invoke it later,
+    /// before moving on to the true completed state.
     ///
-    /// This is only accessed by the timer
-    pub(crate) next_stack: UnsafeCell<Option<Arc<Entry>>>,
-
-    /// Previous entry in the State's linked list.
+    /// There are two reasons this state exists. First, because we cannot take
+    /// the waker after transitioning to a completed state (the [`TimerEntry`]
+    /// could at any time observe this and destroy the [`TimerShared`]); however
+    /// we also can't take the waker immediately before transitioning to a
+    /// completed state, as the future might be in the process of being polled
+    /// and registering a new waker.
     ///
-    /// This is only accessed by the timer and is used to unlink a canceled
-    /// entry.
+    /// To resolve this, we add this state as an intermediate phase where the
+    /// driver can take the waker. If the future is polled at this point, it
+    /// will contend on the driver lock; once it acquires and releases the
+    /// driver lock, it can reliably observe the final completed timer state.
     ///
-    /// This is a weak reference.
-    pub(crate) prev_stack: UnsafeCell<*const Entry>,
+    /// The second reason this exists is that, in general, when we want to
+    /// advance the 'elapsed' time, we might need to process a slot first; when
+    /// doing so, we need to _atomically_ move all of the timers in that slot
+    /// into their new positions, before dropping the lock. However, during this
+    /// time some timers might also need to be fired. If there are too many
+    /// timers awaiting firing to collect their wakers while holding the lock,
+    /// we queue them onto a separate linked list. This state indicates that we
+    /// need to remove the timer from that pending list instead of a timer wheel
+    /// slot.
+    FiringInProgress,
+    /// This timer has been fired due to its timeout being reached. This state
+    /// can be transitioned back to NotRegistered, either by an explicit reset()
+    /// call, or by implicitly detecting a race due to cached and true timeouts
+    /// not matching.
+    Fired,
+    /// This timer has been cancelled. This state exists only on timers in the
+    /// process of being dropped.
+    Cancelled,
+    /// An error has occurred. This state cannot be transitioned out of.
+    Error(crate::time::error::Kind),
 }
 
-/// Stores the info for `Sleep`.
-#[derive(Debug)]
-pub(crate) struct Time {
-    pub(crate) deadline: Instant,
-    pub(crate) duration: Duration,
+impl EntryState {
+    fn is_elapsed(self) -> bool {
+        !matches!(
+            self,
+            EntryState::NotRegistered | EntryState::Registered | EntryState::FiringInProgress
+        )
+    }
 }
 
-/// Flag indicating a timer entry has elapsed
-const ELAPSED: u64 = 1 << 63;
+impl EntryState {
+    fn to_repr(self) -> u8 {
+        unsafe { std::mem::transmute(self) }
+    }
```

Comment:
hmm, this seems to rely on the `ErrorKind` enum being niche-optimized correctly into the `EntryState` enum, doesn't it? if we expect that this is something we can rely on, it might at least be worth having some tests that there are no collisions. and, we might also want to assert that `size_of::<EntryState>() == size_of::<u8>()`, since we can't actually put a repr attribute on the `EntryState` as it has fields...

Reply:
alternatively, we could just flatten the `ErrorKind` variants into `EntryState` and then be able to stick a repr attribute on it, but that might be a pain. :woman_shrugging:

Reply:
I actually went down this path, but it turns out that:

1. `#[repr(u8)]` doesn't actually enforce the representation being `u8` if you have embedded structure fields: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6831cb87fc5d37e1e9bde98e1457fcdf
2. `std::mem::transmute`, however, _does_ perform that enforcement: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6831cb87fc5d37e1e9bde98e1457fcdf

If it becomes a problem we can indeed flatten it, but that adds a bunch of branching everywhere...

Reply:
fair enough, this is _probably_ fine. it might be worth having a test that'll fail if the compiler behavior ever changes so that the transmute doesn't do what we expect it to?

Reply:
You don't even have a `#[repr(...)]` on the enum. You can't assume anything about its layout without a `#[repr(...)]` annotation. It's also unclear what you mean with the `std::mem::transmute` link. It contains the same as the first link.

Reply:
With the changes suggested by Carl that I'll include in the next round, we're won't be transmuting anymore.